src/strategies/spidStrategy.ts

```diff
@@ -0,0 +1,177 @@
+/**
+ * Builds and configure a Passport strategy to authenticate the proxy to the
+ * different SPID IDPs.
+ */
+import { distanceInWordsToNow, isAfter, subDays } from "date-fns";
+import * as SpidStrategy from "spid-passport";
+import * as x509 from "x509";
+import { SpidUser } from "../types/spidUser";
+import {
+  fetchIdpMetadata,
+  IDPOption,
+  mapIpdMetadata,
+  parseIdpMetadata
+} from "../utils/idpLoader";
+import { log } from "../utils/logger";
+
+const IDP_IDS: { [key: string]: string | undefined } = {
+  "https://id.lepida.it/idp/shibboleth": "lepidaid",
+  "https://identity.infocert.it": "infocertid",
+  "https://identity.sieltecloud.it": "sielteid",
+  "https://idp.namirialtsp.com/idp": "namirialid",
+  "https://login.id.tim.it/affwebservices/public/saml2sso": "timid",
+  "https://loginspid.aruba.it": "arubaid",
+  "https://posteid.poste.it": "posteid",
+  "https://spid.intesa.it": "intesaid",
+  "https://spid.register.it": "spiditalia"
+};
+
+/**
+ * Load idp Metadata from a remote url, parse infomations and return a mapped and whitelisted idp options
+ * for spidStrategy object.
+ */
+export async function loadFromRemote(
+  idpMetadataUrl: string
+): Promise<{ [key: string]: IDPOption | undefined }> {
+  log.info("Fetching SPID metadata from [%s]...", idpMetadataUrl);
+  const idpMetadataXML = await fetchIdpMetadata(idpMetadataUrl);
+  log.info("Parsing SPID metadata...");
+  const idpMetadata = parseIdpMetadata(idpMetadataXML);
+  if (idpMetadata.length < Object.keys(IDP_IDS).length) {
+    log.warn("Missing SPID metadata on [%s]", idpMetadataUrl);
+  }
+  log.info("Configuring IdPs...");
+  return mapIpdMetadata(idpMetadata, IDP_IDS);
+}
+
+const spidStrategy = async (spidStrategyConfiguration: {
+  samlKey: string;
+  samlCert: string;
+  samlCallbackUrl: string;
+  samlIssuer: string;
+  samlAcceptedClockSkewMs: number;
+  samlAttributeConsumingServiceIndex: number;
+  spidAutologin: string;
+  spidTestEnvUrl: string;
+  idpMetadataUrl: string;
+}) => {
+  const {
+    samlKey,
+    samlCert,
+    samlCallbackUrl,
+    samlIssuer,
+    samlAcceptedClockSkewMs,
+    samlAttributeConsumingServiceIndex,
+    spidAutologin,
+    spidTestEnvUrl,
+    idpMetadataUrl
+  } = spidStrategyConfiguration;
+  const idpsMetadataOption = await loadFromRemote(idpMetadataUrl);
+
+  logSamlCertExpiration(samlCert);
+
+  const options: {
+    idp: { [key: string]: IDPOption | undefined };
+    // tslint:disable-next-line:no-any
+    sp: any;
+  } = {
+    idp: {
+      ...idpsMetadataOption,
+      xx_testenv2: {
+        cert: [
+          "MIICljCCAX4CCQD95d+Rc57oJzANBgkqhkiG9w0BAQsFADANMQswCQYDVQQGEwJJ" +
```

Comment:
this is for the spid testenv used by io-backend, not sure which one you're going to use

Reply:
I'm not sure I understood: this value comes from [this certificate](https://github.com/teamdigitale/io-onboarding-pa/pull/9/files/fc157d763b95a31b22251c1079db6fd9b8abb595#diff-8f495e4cda4555771ef704dc2a5923d0) which I newly generated. Should I use the one already existing for io-backend instead?